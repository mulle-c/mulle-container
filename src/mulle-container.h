//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
