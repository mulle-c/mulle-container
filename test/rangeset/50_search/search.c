#include <mulle-container/mulle-container.h>

#include <stdio.h>


int   main( int argc, char *argv[])
{
   struct mulle__rangeset   set;
   int                      mode;
   uintptr_t                location;
   uintptr_t                i;

   _mulle__rangeset_init( &set, 0, NULL);

   _mulle__rangeset_insert( &set, mulle_range_make( 2, 2), NULL);
   _mulle__rangeset_insert( &set, mulle_range_make( 6, 2), NULL);

   _mulle__rangeset_print( &set);
   printf( "| # ");
   for( mode = mulle_rangeset_equal; mode <= mulle_rangeset_greater_than_or_equal; mode++)
      printf( "| %s ", mulle_rangeset_searchoperation_utf8_string( mode));
   printf( "|\n");

   printf( "|---");
   for( mode = mulle_rangeset_equal; mode <= mulle_rangeset_greater_than_or_equal; mode++)
      printf( "|%.*s", (int) strlen( mulle_rangeset_searchoperation_utf8_string( mode)) + 2, "--------");
   printf( "|\n");

   for( i = 0; i < 10; i++)
   {
      printf( "| %tu ", i);
      for( mode = mulle_rangeset_equal; mode <= mulle_rangeset_greater_than_or_equal; mode++)
      {
         location = _mulle__rangeset_search_location( &set, i, mode);
         printf( "|%.*s", (int) strlen( mulle_rangeset_searchoperation_utf8_string( mode)), "        ");
         if( location  == mulle_not_found_e)
            printf( "- ");
         else
            printf( "%tu ", location);
      }
      printf( "|\n");
   }
   _mulle__rangeset_done( &set, NULL);

   return( 0);
}
