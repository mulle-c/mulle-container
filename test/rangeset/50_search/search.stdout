#0: 2,2
#1: 6,2
| # | == | < | <= | > | >= |
|---|----|---|----|---|----|
| 0 |  - | - |  - | 2 |  2 |
| 1 |  - | - |  - | 2 |  2 |
| 2 |  2 | - |  2 | 3 |  2 |
| 3 |  3 | 2 |  3 | 6 |  3 |
| 4 |  - | 3 |  3 | 6 |  6 |
| 5 |  - | 3 |  3 | 6 |  6 |
| 6 |  6 | 3 |  6 | 7 |  6 |
| 7 |  7 | 6 |  7 | - |  7 |
| 8 |  - | 7 |  7 | - |  - |
| 9 |  - | 7 |  7 | - |  - |
